// @chakra-ui/react/dist/cjs/descendant/use-descendant.cjs
'use client';
'use strict';

var context = require('@chakra-ui/utils/context');
var useMergeRefs = require('@chakra-ui/hooks/use-merge-refs');
var react = require('react');
var descendant = require('./descendant.cjs');
var utils = require('./utils.cjs');

function useDescendants() {
  const descendants = react.useRef(new descendant.DescendantsManager());
  utils.useSafeLayoutEffect(() => {
    return () => descendants.current.destroy();
  });
  return descendants.current;
}
const [DescendantsContextProvider, useDescendantsContext] = context.createContext({
  name: "DescendantsProvider",
  errorMessage: "useDescendantsContext must be used within DescendantsProvider"
});
function useDescendant(options) {
  const descendants = useDescendantsContext();
  const [index, setIndex] = react.useState(-1);
  const ref = react.useRef(null);
  utils.useSafeLayoutEffect(() => {
    return () => {
      if (!ref.current)
        return;
      descendants.unregister(ref.current);
    };
  }, []);
  utils.useSafeLayoutEffect(() => {
    if (!ref.current)
      return;
    const dataIndex = Number(ref.current.dataset["index"]);
    if (index != dataIndex && !Number.isNaN(dataIndex)) {
      setIndex(dataIndex);
    }
  });
  const refCallback = options ? utils.cast(descendants.register(options)) : utils.cast(descendants.register);
  return {
    descendants,
    index,
    enabledIndex: descendants.enabledIndexOf(ref.current),
    register: useMergeRefs.mergeRefs(refCallback, ref)
  };
}
function createDescendantContext() {
  const ContextProvider = utils.cast(DescendantsContextProvider);
  const _useDescendantsContext = () => utils.cast(useDescendantsContext());
  const _useDescendant = (options) => useDescendant(options);
  const _useDescendants = () => useDescendants();
  return [
    // context provider
    ContextProvider,
    // call this when you need to read from context
    _useDescendantsContext,
    // descendants state information, to be called and passed to `ContextProvider`
    _useDescendants,
    // descendant index information
    _useDescendant
  ];
}

exports.createDescendantContext = createDescendantContext;
