// @chakra-ui/react/dist/cjs/accordion/accordion-context.cjs
'use client';
'use strict';

var context = require('@chakra-ui/utils/context');
var useDescendant = require('../descendant/use-descendant.cjs');

const [AccordionStylesProvider, useAccordionStyles] = context.createContext({
  name: "AccordionStylesContext",
  hookName: "useAccordionStyles",
  providerName: "<Accordion />"
});
const [AccordionItemProvider, useAccordionItemContext] = context.createContext({
  name: "AccordionItemContext",
  hookName: "useAccordionItemContext",
  providerName: "<AccordionItem />"
});
const [
  AccordionDescendantsProvider,
  useAccordionDescendantsContext,
  useAccordionDescendants,
  useAccordionDescendant
] = useDescendant.createDescendantContext();

exports.AccordionDescendantsProvider = AccordionDescendantsProvider;
exports.AccordionItemProvider = AccordionItemProvider;
exports.AccordionStylesProvider = AccordionStylesProvider;
exports.useAccordionDescendant = useAccordionDescendant;
exports.useAccordionDescendants = useAccordionDescendants;
exports.useAccordionDescendantsContext = useAccordionDescendantsContext;
exports.useAccordionItemContext = useAccordionItemContext;
exports.useAccordionStyles = useAccordionStyles;
