// @chakra-ui/react/dist/cjs/grid/grid-item.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var compact = require('@chakra-ui/utils/compact');
var responsive = require('@chakra-ui/utils/responsive');
var forwardRef = require('../system/forward-ref.cjs');
var factory = require('../system/factory.cjs');

function spanFn(span) {
  return responsive.mapResponsive(
    span,
    (value) => value === "auto" ? "auto" : `span ${value}/span ${value}`
  );
}
const GridItem = forwardRef.forwardRef(
  function GridItem2(props, ref) {
    const {
      area,
      colSpan,
      colStart,
      colEnd,
      rowEnd,
      rowSpan,
      rowStart,
      ...rest
    } = props;
    const styles = compact.compact({
      gridArea: area,
      gridColumn: spanFn(colSpan),
      gridRow: spanFn(rowSpan),
      gridColumnStart: colStart,
      gridColumnEnd: colEnd,
      gridRowStart: rowStart,
      gridRowEnd: rowEnd
    });
    return /* @__PURE__ */ jsxRuntime.jsx(factory.chakra.div, { ref, __css: styles, ...rest });
  }
);
GridItem.displayName = "GridItem";

exports.GridItem = GridItem;
